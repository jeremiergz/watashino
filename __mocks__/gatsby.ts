import Gatsby from 'gatsby';

module.exports = {
  ...Gatsby,
  graphql: jest.fn(),
  navigate: jest.fn(),
  StaticQuery: jest.fn(),
  useStaticQuery: jest.fn(),
};
