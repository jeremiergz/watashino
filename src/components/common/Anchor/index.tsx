import React from 'react';
import styled from 'styled-components';
import { variant } from 'styled-system';
import BaseBox, { BoxProps } from '../../primitives/Box';

type AnchorProps = BoxProps & {
  download?: boolean;
  href?: string;
  target?: string;
  variant?: 'underlined';
};

const A = styled(BaseBox)`
  color: inherit;
  text-decoration: none;
  transition: color 100ms ease-in-out;
  .gatsby-image-wrapper {
    transition: filter 100ms ease-in-out;
  }
  &:hover {
    color: ${({ theme }) => theme.colors.primary};
    .gatsby-image-wrapper {
      filter: brightness(0.75);
    }
  }
  ${variant({
    variants: {
      underlined: {
        textDecoration: 'underline',
      },
    },
  })}
`;

const Anchor = ({ children, ...rest }: AnchorProps) => {
  const isTouchDevice = typeof window !== 'undefined' && window.navigator.maxTouchPoints > 0;
  const target = isTouchDevice ? '_self' : '_blank';
  return (
    <A as="a" fontWeight="semi-bold" target={target} {...rest}>
      {children}
    </A>
  );
};

Anchor.displayName = 'Anchor';

export { AnchorProps };
export default Anchor;
