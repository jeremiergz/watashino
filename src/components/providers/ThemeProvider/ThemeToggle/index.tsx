import Button from 'components/Button';
import FlexBox from 'components/primitives/FlexBox';
import Text from 'components/primitives/Text';
import MoonIcon from 'components/svg/Moon';
import SunIcon from 'components/svg/Sun';
import React from 'react';
import { useTheming } from '..';

const iconMapping = {
  dark: MoonIcon,
  light: SunIcon,
};

const ThemeToggle: React.FC = () => {
  const { theme, toggle } = useTheming();
  const ThemeIcon = iconMapping[theme.type];
  return (
    <Button onClick={() => toggle()} position="absolute" right={3} top={3} variant="outlined" width={72}>
      <FlexBox alignItems="center" justifyContent="space-between" width="100%">
        <ThemeIcon height={20} width={20} />
        <Text color="gray" fontFamily={theme.fonts.main} fontSize={14}>
          {theme.type}
        </Text>
      </FlexBox>
    </Button>
  );
};

ThemeToggle.displayName = 'ThemeToggle';

export default ThemeToggle;
