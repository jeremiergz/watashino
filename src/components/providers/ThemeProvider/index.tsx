import React, { createContext, PropsWithChildren, useContext, useEffect, useState } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { DarkTheme, LightTheme, Theme } from 'theme';

const defaultMode = 'light';
const localStorageKey = 'theme:mode';

const ThemeContext = createContext<{ mode: ThemeMode; theme: Theme; toggle: () => void }>({
  mode: defaultMode,
  theme: LightTheme,
  toggle: () => null,
});

const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
  const isBrowser = typeof window !== 'undefined';
  const [isInitialized, setIsInitialized] = useState(false);
  const [mode, setMode] = useState<ThemeMode>(defaultMode);
  const toggle = () => {
    const newMode = mode === 'light' ? 'dark' : 'light';
    setMode(newMode);
  };
  const theme = mode === 'light' ? LightTheme : DarkTheme;
  useEffect(() => {
    if (isBrowser && isInitialized) localStorage.setItem(localStorageKey, mode);
  }, [mode]);
  useEffect(() => {
    const storedMode = (isBrowser && (localStorage.getItem(localStorageKey) as ThemeMode)) || defaultMode;
    setMode(storedMode);
    setIsInitialized(true);
  }, []);
  return (
    <ThemeContext.Provider value={{ mode, theme, toggle }}>
      <StyledThemeProvider theme={theme}>{children}</StyledThemeProvider>
    </ThemeContext.Provider>
  );
};

ThemeProvider.displayName = 'ThemeProvider';

/**
 * Returns the theming context that allows retrieval of theme & toggling theme mode.
 */
function useTheming() {
  return useContext(ThemeContext);
}

export { useTheming };
export type ThemeProviderProps = PropsWithChildren<{}>;
export default ThemeProvider;
