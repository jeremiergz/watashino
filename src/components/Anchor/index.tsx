import Box, { BoxProps } from 'components/primitives/Box';
import { Theme } from 'components/providers/ThemeProvider';
import React from 'react';
import styled from 'styled-components';
import { variant } from 'styled-system';

const A = styled(Box)`
  color: inherit;
  text-decoration: none;
  transition: color 100ms ease-in-out, filter 100ms ease-in-out;
  will-change: filter;
  &:hover {
    color: ${({ theme }: { theme: Theme }) => theme.colors.primary};
    filter: brightness(0.75);
  }
  ${variant({
    variants: {
      underlined: {
        textDecoration: 'underline',
      },
    },
  })}
`;

const Anchor: React.FC<AnchorProps> = ({ children, ...rest }) => {
  const isTouchDevice = typeof window !== 'undefined' && window.navigator.maxTouchPoints > 0;
  const target = isTouchDevice ? '_self' : '_blank';
  return (
    <A as="a" fontWeight="semiBold" target={target} {...rest}>
      {children}
    </A>
  );
};

Anchor.displayName = 'Anchor';

export type AnchorProps = BoxProps & {
  href?: string;
  target?: string;
  variant?: 'underlined';
};
export default Anchor;
