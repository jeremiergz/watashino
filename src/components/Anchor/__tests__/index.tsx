import ThemeProvider from 'components/providers/ThemeProvider';
import * as Gatsby from 'gatsby';
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Anchor from '..';

describe('components/Anchor Test Suite', () => {
  beforeAll(() => {
    const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');
    useStaticQuery.mockImplementation(() => ({
      themeJson: {
        colors: {
          black: '#212121',
          dark: '#424242',
          error: '#f44336',
          gray: '#757575',
          info: '#2196f3',
          lightgray: '#e0e0e0',
          primary: '#31859a',
          secondary: '#17365c',
          success: '#4caf50',
          vsCodeDark: '#1e1e1e',
          white: '#fafafa',
        },
        fonts: {
          main: "'Open Sans', sans-serif",
        },
        fontWeights: {
          bold: 800,
          regular: 400,
          semiBold: 600,
        },
      },
    }));
  });

  it('matches snapshot', () => {
    const jsx = (
      <ThemeProvider>
        <Anchor variant="underlined">Anchor Component</Anchor>
      </ThemeProvider>
    );
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });

  it('should use target="_blank" on desktop devices by default', () => {
    Object.assign(window.navigator, { maxTouchPoints: 0 });
    const jsx = (
      <ThemeProvider>
        <Anchor variant="underlined">Anchor Component</Anchor>
      </ThemeProvider>
    );
    const renderedInstance = ReactTestRenderer.create(jsx).root;
    expect(renderedInstance.findByType('a').props.target).toBe('_blank');
  });

  it('should use target="_self" on touch devices by default', () => {
    Object.assign(window.navigator, { maxTouchPoints: 4 });
    const jsx = (
      <ThemeProvider>
        <Anchor variant="underlined">Anchor Component</Anchor>
      </ThemeProvider>
    );
    const renderedInstance = ReactTestRenderer.create(jsx).root;
    expect(renderedInstance.findByType('a').props.target).toBe('_self');
  });
});
