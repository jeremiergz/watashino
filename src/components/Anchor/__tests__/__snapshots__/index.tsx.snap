// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Anchor Test Suite matches snapshot 1`] = `
<a
  className="sc-AxjAm sc-AxirZ dWhqPI"
  fontWeight="semiBold"
  target="_blank"
>
  Anchor Component
</a>
`;
